src/orchestration/Orchestration.ts

```diff
@@ -415,16 +415,21 @@ export class Orchestration {
             this.config
         );
 
-        if (this.config.identityPoolId && this.config.guestRoleArn) {
-            dispatch.setAwsCredentials(
-                new Authentication(this.config)
-                    .ChainAnonymousCredentialsProvider
-            );
-        } else if (this.config.identityPoolId) {
-            dispatch.setAwsCredentials(
-                new EnhancedAuthentication(this.config)
-                    .ChainAnonymousCredentialsProvider
-            );
+        // Only retrieves and sets credentials if the session is sampled.
+        // The first nil session will have the same sampling decision as the
+        // new session created when the first event is recorded.
+        if (this.eventCache.isSessionSampled()) {
```

Comment:
nit: we can perhaps just do `return dispatch;` if the session isn't sampled to avoid nested if else

Reply:
Refactored!